cli_tools/gce_ovf_export/exporter/steps.go

```diff
@@ -0,0 +1,212 @@
+type populateStepsFunc func(*daisy.Workflow) error
+
+func (oe *OVFExporter) prepare(ctx context.Context, instance *compute.Instance) error {
+	return oe.runStep(ctx, func() error {
+		oe.Logger.Log(fmt.Sprintf("Stopping '%v' instance and detaching the disks.", instance.Name))
+		return oe.instanceExportPreparer.Prepare(instance, oe.params)
+	}, oe.instanceExportPreparer.Cancel, oe.instanceExportPreparer.TraceLogs)
+}
+
+func (oe *OVFExporter) exportDisks(ctx context.Context, instance *compute.Instance) error {
+	return oe.runStep(ctx, func() error {
+		oe.Logger.Log("Exporting the disks.")
+		var err error
+		oe.exportedDisks, err = oe.instanceDisksExporter.Export(instance, oe.params)
+		return err
+	}, oe.instanceDisksExporter.Cancel, oe.instanceDisksExporter.TraceLogs)
+}
+
+func (oe *OVFExporter) inspectBootDisk(ctx context.Context) error {
+	return oe.runStep(ctx, func() error {
+		oe.Logger.Log("Inspecting the boot disk.")
+		bootDisk := getBootDisk(oe.exportedDisks)
+		if bootDisk == nil {
+			return nil
+		}
+		var err error
+		oe.bootDiskInspectionResults, err = oe.inspector.Inspect(
+			daisyutils.GetDiskURI(*oe.params.Project, oe.params.Zone, bootDisk.Disk.Name), true)
+		if err != nil {
+			oe.Logger.Log(fmt.Sprintf("WARNING: Could not detect operating system on the boot disk: %v", err))
+		}
+		oe.Logger.Log(fmt.Sprintf("Disk inspection results: %v", oe.bootDiskInspectionResults))
+		// don't return error if inspection fails, just log it, since it's not a show-stopper.
+		return nil
+	}, oe.inspector.Cancel, oe.inspector.TraceLogs)
+}
+
+func (oe *OVFExporter) generateDescriptor(ctx context.Context, instance *compute.Instance) error {
+	return oe.runStep(ctx, func() error {
+		oe.Logger.Log("Generating OVF descriptor.")
+		bucketName, gcsDirectoryPath, err := storageutils.GetGCSObjectPathElements(oe.params.DestinationURI)
+		if err != nil {
+			return err
+		}
+		return oe.ovfDescriptorGenerator.GenerateAndWriteOVFDescriptor(instance, oe.exportedDisks, bucketName, gcsDirectoryPath, oe.bootDiskInspectionResults)
+	}, oe.ovfDescriptorGenerator.Cancel, func() []string { return nil })
+}
+
+func (oe *OVFExporter) generateManifest(ctx context.Context) error {
+	return oe.runStep(ctx, func() error {
+		oe.Logger.Log("Generating manifest.")
+		return oe.manifestFileGenerator.GenerateAndWriteToGCS(oe.params.DestinationURI, oe.params.InstanceName)
+	}, oe.manifestFileGenerator.Cancel, func() []string { return nil })
+}
+
+func (oe *OVFExporter) cleanup(instance *compute.Instance, exportError error) error {
+	// cleanup shouldn't react to time out as it's necessary to perform this step.
+	// Otherwise, instance being exported would be left shut down and disks detached.
+	oe.Logger.Log("Cleaning up.")
+	if err := oe.instanceExportCleaner.Clean(instance, oe.params); err != nil {
+		return err
+	}
+	if exportError == nil {
+		oe.Logger.Log("OVF export finished successfully.")
+	}
+	if oe.storageClient != nil {
+		err := oe.storageClient.Close()
```

Comment:
Why close here? 1. is it required to close? 2. what if it's not closed (for example, exportError != nil)?

Reply:
If there's an error, the export is still done and we close storage client as we will not writing any files (OVF descriptor, disk files) to GCS. We always close storage client once it's not needed.